README.md
# CERNBox Group Daemon

The CERNBox group daemon is a dameon that caches LDAP information from central
CERN LDAP for finding:

- all the egroups a user belongs to 
- all the users inside an egroup

Redis is used as shared cache allowing multiple instances of cboxgroupd to be
run simultaneously

## Options

```
Usage of cboxgroupd:
  -applog string
        File to log application data (default "stderr")
  -httplog string
        File to log HTTP requests (default "stderr")
  -ldaphostname string
        Hostname of the LDAP server (default "xldap.cern.ch")
  -ldappagelimit uint
        Page limit for paged searchs (default 1000)
  -ldapport int
        Port of LDAP server (default 389)
  -port int
        Port to listen for connections (default 2002)
  -redisdb int
        Redis number database for keys isolation (0-15)
  -redishostname string
        Hostname of the Redis server (default "localhost")
  -redisport int
        Port of Redis server (default 6379)
  -redisttl int
        Number of seconds to expire cached entries in Redis (default 60)
  -secret string
        Share secret between services to authenticate requests (default "changeme!!!")
  -version
        Show version

```

## How to run it

```
$ cboxgroupd --secret abc --port 2002

{"level":"info","ts":1497533157.7921486,"caller":"/home/labkode/go/src/github.com/cernbox/cboxgroupd/main.go:81","msg":"server is listening","port":3000}
....
```

## Some example requests

```
curl -i localhost:2002/api/v1/membership/usersingroup/cernbox-admins -H "Authorization: Bearer abc"

curl -i localhost:2002/api/v1/membership/usersincomputinggroup/zp -H "Authorization: Bearer abc"

curl -i localhost:2002/api/v1/membership/usergroups/gonzalhu -H "Authorization: Bearer abc"

curl -i localhost:2002/api/v1/search/hugo -H "Authorization: Bearer abc" (searchs for primary users, egroups and unix groups)

curl -i localhost:2002/api/v1/search/a:labrador -H "Authorization: Bearer abc" (searchs for all users accounts, egroups and unix groups)

curl -i localhost:2002/api/v1/search/g:def-cg -H "Authorization: Bearer abc" (search for unix groups)

```



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.4.0] - 2017-11-22
### Added
- New endpoint to perform searchs (*/api/v1/search/{filter}*)


pkg/ldapgrouplooker/ldapgrouplooker_test.go
package ldapgrouplooker

import (
	"context"
	"fmt"
	"testing"
)

func TestUserGroups(t *testing.T) {
	ctx := context.Background()
	gl := New("xldap.cern.ch", 389, 1000)
	gids, err := gl.GetUserGroups(ctx, "gonzalhu", false)
	if err != nil {
		t.Error(err)
	}
	for _, g := range gids {
		fmt.Println(g)
	}
}

func TestComputingGroups(t *testing.T) {
	ctx := context.Background()
	gl := New("xldap.cern.ch", 389, 1000)
	gids, err := gl.GetUserComputingGroups(ctx, "gonzalhu", false)
	if err != nil {
		t.Error(err)
	}
	for _, g := range gids {
		fmt.Println(g)
	}
}


pkg/pkg.go
package pkg

import (
	"context"
	"fmt"
	"time"
)

type LDAPAccountType string

var (
	LDAPAccountTypePrimary   LDAPAccountType = "primary"
	LDAPAccountTypeSecondary LDAPAccountType = "secondary"
	LDAPAccountTypeService   LDAPAccountType = "service"
	LDAPAccountTypeEGroup    LDAPAccountType = "egroup"
	LDAPAccountTypeUnixGroup LDAPAccountType = "unixgroup"
	LDAPAccountTypeUndefined LDAPAccountType = "undefined"
)

type SearchEntry struct {
	DN          string          `json:"dn"`
	CN          string          `json:"cn"`
	AccountType LDAPAccountType `json:"account_type"`
	DisplayName string          `json:"display_name"`
	Mail        string          `json:"mail"`
}

type GroupLookerErrorCode string

const (
	GroupLookerErrorNotFound GroupLookerErrorCode = "GROUPLOOKER_ERROR_NOT_FOUND"
)

func NewGroupLookerError(code GroupLookerErrorCode) GroupLookerError {
	return GroupLookerError{Code: code}
}

type GroupLookerError struct {
	Code    GroupLookerErrorCode
	Message string
}

func (sr GroupLookerError) WithMessage(msg string) GroupLookerError {
	sr.Message = msg
	return sr
}

func (sr GroupLookerError) Error() string {
	return fmt.Sprintf("%s: %s", sr.Code, sr.Message)
}

type GroupLooker interface {
	GetUsersInGroup(ctx context.Context, gid string, cached bool) ([]string, error)
	GetUserGroups(ctx context.Context, uid string, cached bool) ([]string, error)
	GetUsersInComputingGroup(ctx context.Context, gid string, cached bool) ([]string, error)
	GetUserComputingGroups(ctx context.Context, gid string, cached bool) ([]string, error)
	GetTTLForUser(ctx context.Context, uid string) (time.Duration, error)
	GetTTLForGroup(ctx context.Context, gid string) (time.Duration, error)
	GetTTLForComputingUser(ctx context.Context, uid string) (time.Duration, error)
	GetTTLForComputingGroup(ctx context.Context, gid string) (time.Duration, error)
	Search(ctx context.Context, filter string, cached bool) ([]*SearchEntry, error)
}


create_releases.sh
#!/usr/bin/env bash
#
# ClawIO build script. Add build variables to compiled binary.
# # Usage:
#
#     $ ./build.bash [output_filename] [git_repo]
#
# Outputs compiled program in current directory.
# Default file name is 'clawiod'.
# Default git repo is current directory.
# Builds always take place from current directory.

set -euo pipefail

: ${output_filename:="${1:-}"}
: ${output_filename:="cboxgroupd"}

: ${git_repo:="${2:-}"}
: ${git_repo:="."}

pkg=main
ldflags=()

# Timestamp of build
ts_name="${pkg}.buildDate"
ts_value=$(date -u +"%a %b %d %H:%M:%S %Z %Y")
ldflags+=("-X" "\"${ts_name}=${ts_value}\"")

# Current tag, if HEAD is on a tag
# This value is used to determine if the current build is a dev build or a release build
# If this value is empty means we are not on an tag, thus is a dev build
current_tag_name="${pkg}.gitTag"
set +e
current_tag_value="$(git -C "${git_repo}" describe --exact-match HEAD 2>/dev/null)"
set -e
ldflags+=("-X" "\"${current_tag_name}=${current_tag_value}\"")

# Nearest tag on branch
tag_name="${pkg}.gitNearestTag"
tag_value="$(git -C "${git_repo}" describe --abbrev=0 --tags HEAD)"
ldflags+=("-X" "\"${tag_name}=${tag_value}\"")

# Commit SHA
commit_name="${pkg}.gitCommit"
commit_value="$(git -C "${git_repo}" rev-parse --short HEAD)"
ldflags+=("-X" "\"${commit_name}=${commit_value}\"")

# Application name
app_name="${pkg}.appName"
ldflags+=("-X" "\"${app_name}=${output_filename}\"")


releases_dir=${git_repo}/releases
rm -rf ${releases_dir}
mkdir -p ${releases_dir}

os=( "linux" "darwin" "windows" )
arch=( "amd64")

if [[ -z "${current_tag_value}" ]]; then
	# dev build
	current_date=$(date +"%Y%m%d%H%M%S")
	for i in "${os[@]}"; do
		for j in "${arch[@]}"; do
			GOOS=$i GOARCH=$j go build -ldflags "${ldflags[*]}" -o ${releases_dir}/"${output_filename}"-${tag_value}-$i-$j-${current_date}-${commit_value}
		done;
	done;
else
	# release build
	for i in "${os[@]}"; do
		for j in "${arch[@]}"; do
			artifact_folder=${releases_dir}/"${output_filename}"-${tag_value}-$i-$j
			mkdir ${artifact_folder}
			GOOS=$i GOARCH=$j go build -ldflags "${ldflags[*]}" -o ${artifact_folder}/"${output_filename}"
			cp LICENSE ${artifact_folder}
			tar -cvzf "${artifact_folder}".tar.gz "${artifact_folder}"
		done;
	done;
fi


handlers/handlers.go
package handlers

import (
	"encoding/json"
	"github.com/cernbox/cboxgroupd/pkg"
	"github.com/gorilla/mux"
	"go.uber.org/zap"
	"io/ioutil"
	"net/http"
	"strings"
	"sync"
)

func CheckSharedSecret(logger *zap.Logger, secret string, handler http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// The secret is passed in the header: Authorization: Bearer mysecret
		h := r.Header.Get("Authorization")
		secret := "bearer " + secret
		if secret != strings.ToLower(h) {
			logger.Warn("wrong secret")
			w.WriteHeader(http.StatusUnauthorized)
			return
		}
		handler.ServeHTTP(w, r)
	})
}

func Search(logger *zap.Logger, groupLooker pkg.GroupLooker) http.Handler {

	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		filter := mux.Vars(r)["filter"]
		if filter == "" {
			logger.Error("filter is empty")
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		entries, err := groupLooker.Search(r.Context(), filter, true)
		if err != nil {
			logger.Info("error getting entries", zap.Error(err), zap.String("filter", filter))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
		logger.Info("entries found", zap.Int("numentries", len(entries)), zap.String("filter", filter))
		json.NewEncoder(w).Encode(entries)
	})
}

func UsersInGroup(logger *zap.Logger, groupLooker pkg.GroupLooker) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gid := mux.Vars(r)["gid"]
		if gid == "" {
			logger.Error("gid is empty")
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		uids, err := groupLooker.GetUsersInGroup(r.Context(), gid, true)
		if err != nil {
			if gle, ok := err.(pkg.GroupLookerError); ok {
				if gle.Code == pkg.GroupLookerErrorNotFound {
					logger.Warn("group not found", zap.String("gid", gid))
					w.WriteHeader(http.StatusNotFound)
					return
				}
			}
			logger.Info("error getting users", zap.Error(err), zap.String("gid", gid))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
		logger.Info("users found", zap.Int("numusers", len(uids)), zap.String("gid", gid))
		json.NewEncoder(w).Encode(uids)
	})
}

func UsersInComputingGroup(logger *zap.Logger, groupLooker pkg.GroupLooker) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gid := mux.Vars(r)["gid"]
		if gid == "" {
			logger.Error("gid is empty")
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		uids, err := groupLooker.GetUsersInComputingGroup(r.Context(), gid, true)
		if err != nil {
			if gle, ok := err.(pkg.GroupLookerError); ok {
				if gle.Code == pkg.GroupLookerErrorNotFound {
					logger.Warn("computing group not found", zap.String("gid", gid))
					w.WriteHeader(http.StatusNotFound)
					return
				}
			}
			logger.Info("error getting users", zap.Error(err), zap.String("gid", gid))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
		logger.Info("users found", zap.Int("numusers", len(uids)), zap.String("gid", gid))
		json.NewEncoder(w).Encode(uids)
	})
}

func UserGroups(logger *zap.Logger, groupLooker pkg.GroupLooker) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		uid := mux.Vars(r)["uid"]
		if uid == "" {
			logger.Error("uid is empty")
			w.WriteHeader(http.StatusBadRequest)
			return
		}
		gids, err := groupLooker.GetUserGroups(r.Context(), uid, true)
		if err != nil {
			if gle, ok := err.(pkg.GroupLookerError); ok {
				if gle.Code == pkg.GroupLookerErrorNotFound {
					logger.Warn("user not found", zap.String("uid", uid))
					w.WriteHeader(http.StatusNotFound)
					return
				}
			}
			logger.Info("error getting users", zap.Error(err), zap.String("uid", uid))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
		logger.Info("groups found", zap.Int("numgroups", len(gids)), zap.String("uid", uid))
		json.NewEncoder(w).Encode(gids)
	})
}

func UserComputingGroups(logger *zap.Logger, groupLooker pkg.GroupLooker) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		uid := mux.Vars(r)["uid"]
		if uid == "" {
			logger.Error("uid is empty")
			w.WriteHeader(http.StatusBadRequest)
			return
		}
		gids, err := groupLooker.GetUserComputingGroups(r.Context(), uid, true)
		if err != nil {
			if gle, ok := err.(pkg.GroupLookerError); ok {
				if gle.Code == pkg.GroupLookerErrorNotFound {
					logger.Warn("user not found", zap.String("uid", uid))
					w.WriteHeader(http.StatusNotFound)
					return
				}
			}
			logger.Info("error getting users", zap.Error(err), zap.String("uid", uid))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
		logger.Info("unix groups found", zap.Int("numgroups", len(gids)), zap.String("uid", uid))
		json.NewEncoder(w).Encode(gids)
	})
}

func UserGroupsTTL(logger *zap.Logger, groupLooker pkg.GroupLooker) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		uid := mux.Vars(r)["uid"]
		if uid == "" {
			logger.Error("uid is empty")
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		ttl, err := groupLooker.GetTTLForUser(r.Context(), uid)
		if err != nil {
			logger.Info("error getting ttl for user", zap.Error(err), zap.String("uid", uid))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}

		res := struct {
			UID string  `json:"uid"`
			TTL float64 `json:"ttl"`
		}{uid, ttl.Seconds()}
		logger.Info("ttl retrieved", zap.String("uid", uid), zap.Float64("ttl", res.TTL))
		json.NewEncoder(w).Encode(res)
	})
}

func UserComputingGroupsTTL(logger *zap.Logger, groupLooker pkg.GroupLooker) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		uid := mux.Vars(r)["uid"]
		if uid == "" {
			logger.Error("uid is empty")
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		ttl, err := groupLooker.GetTTLForComputingUser(r.Context(), uid)
		if err != nil {
			logger.Info("error getting ttl for computing user", zap.Error(err), zap.String("uid", uid))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}

		res := struct {
			UID string  `json:"uid"`
			TTL float64 `json:"ttl"`
		}{uid, ttl.Seconds()}
		logger.Info("ttl retrieved", zap.String("uid", uid), zap.Float64("ttl", res.TTL))
		json.NewEncoder(w).Encode(res)
	})
}

func UsersInGroupTTL(logger *zap.Logger, groupLooker pkg.GroupLooker) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gid := mux.Vars(r)["gid"]
		if gid == "" {
			logger.Error("gid is empty")
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		ttl, err := groupLooker.GetTTLForGroup(r.Context(), gid)
		if err != nil {
			logger.Info("error getting ttl for group", zap.Error(err), zap.String("gid", gid))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}

		res := struct {
			GID string  `json:"gid"`
			TTL float64 `json:"ttl"`
		}{gid, ttl.Seconds()}
		logger.Info("ttl retrieved", zap.String("gid", gid), zap.Float64("ttl", res.TTL))
		json.NewEncoder(w).Encode(res)
	})
}

func UsersInComputingGroupTTL(logger *zap.Logger, groupLooker pkg.GroupLooker) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gid := mux.Vars(r)["gid"]
		if gid == "" {
			logger.Error("gid is empty")
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		ttl, err := groupLooker.GetTTLForComputingGroup(r.Context(), gid)
		if err != nil {
			logger.Info("error getting ttl for computing group", zap.Error(err), zap.String("gid", gid))
			w.WriteHeader(http.StatusInternalServerError)
			return
		}

		res := struct {
			GID string  `json:"gid"`
			TTL float64 `json:"ttl"`
		}{gid, ttl.Seconds()}
		logger.Info("ttl retrieved", zap.String("gid", gid), zap.Float64("ttl", res.TTL))
		json.NewEncoder(w).Encode(res)
	})
}

// UpdateUsersInGroups allows to trigger a refresh of users belonfing to a group
func UpdateUsersInGroup(logger *zap.Logger, groupLooker pkg.GroupLooker, maxConcurrency int) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		type request struct {
			Groups []string `json:"groups"`
		}
		data, err := ioutil.ReadAll(r.Body)
		if err != nil {
			logger.Error(err.Error())
			w.WriteHeader(http.StatusBadRequest)
			return
		}
		req := &request{}
		err = json.Unmarshal(data, req)
		if err != nil {
			logger.Error(err.Error())
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		go func() {
			var throttle = make(chan int, maxConcurrency)
			var wg sync.WaitGroup
			for _, gid := range req.Groups {
				if gid != "" {
					throttle <- 1
					wg.Add(1)
					go func(wg *sync.WaitGroup, throttle chan int, gid string) {
						defer wg.Done()
						defer func() { <-throttle }()
						uids, err := groupLooker.GetUsersInGroup(r.Context(), gid, false)
						if err != nil {
							if gle, ok := err.(pkg.GroupLookerError); ok {
								if gle.Code == pkg.GroupLookerErrorNotFound {
									logger.Warn("async: group not found", zap.String("gid", gid))
									return
								}
							}
							logger.Info("async: error getting users", zap.Error(err), zap.String("gid", gid))
						}
						logger.Info("async: users found", zap.Int("numusers", len(uids)), zap.String("gid", gid))
					}(&wg, throttle, gid)
				}
			}
			wg.Wait()
		}()
		w.WriteHeader(http.StatusAccepted)
	})
}

// UpdateUserGroups allows to trigger a refresh of groups belonfing to an user
func UpdateUserGroups(logger *zap.Logger, groupLooker pkg.GroupLooker, maxConcurrency int) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		type request struct {
			Users []string `json:"users"`
		}
		data, err := ioutil.ReadAll(r.Body)
		if err != nil {
			logger.Error(err.Error())
			w.WriteHeader(http.StatusBadRequest)
			return
		}
		req := &request{}
		err = json.Unmarshal(data, req)
		if err != nil {
			logger.Error(err.Error())
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		go func() {
			var throttle = make(chan int, maxConcurrency)
			var wg sync.WaitGroup
			for _, uid := range req.Users {
				if uid != "" {
					throttle <- 1
					wg.Add(1)
					go func(wg *sync.WaitGroup, throttle chan int, uid string) {
						defer wg.Done()
						defer func() { <-throttle }()
						gids, err := groupLooker.GetUserGroups(r.Context(), uid, false)
						if err != nil {
							if gle, ok := err.(pkg.GroupLookerError); ok {
								if gle.Code == pkg.GroupLookerErrorNotFound {
									logger.Warn("async: user not found", zap.String("uid", uid))
								}
							}
							logger.Info("async: error getting groups", zap.Error(err), zap.String("uid", uid))
							return
						}
						logger.Info("async: groups found", zap.Int("numgroups", len(gids)), zap.String("uid", uid))
					}(&wg, throttle, uid)
				}
			}
			wg.Wait()
		}()
		w.WriteHeader(http.StatusAccepted)
	})
}


main.go
package main

import (
	"flag"
	"fmt"
	"github.com/cernbox/cboxgroupd/handlers"
	"github.com/cernbox/cboxgroupd/pkg/ldapgrouplooker"
	"github.com/cernbox/cboxgroupd/pkg/redisgrouplooker"
	gh "github.com/gorilla/handlers"
	"github.com/gorilla/mux"
	"github.com/spf13/pflag"
	"github.com/spf13/viper"
	"go.uber.org/zap"
	"log"
	"net/http"
	"os"
)

// Build information obtained with the help of -ldflags
var (
	appName       string
	buildDate     string // date -u
	gitTag        string // git describe --exact-match HEAD
	gitNearestTag string // git describe --abbrev=0 --tags HEAD
	gitCommit     string // git rev-parse HEAD
)

var fVersion bool

func init() {
	viper.SetDefault("port", 2002)
	viper.SetDefault("ldaphostname", "xldap.cern.ch")
	viper.SetDefault("ldapport", 389)
	viper.SetDefault("ldappagelimit", 1000)
	viper.SetDefault("redishostname", "localhost")
	viper.SetDefault("redisport", 6379)
	viper.SetDefault("redisdb", 0)
	viper.SetDefault("redisttl", 60)
	viper.SetDefault("applog", "stderr")
	viper.SetDefault("httplog", "stderr")
	viper.SetDefault("secret", "change_me!!!")
	viper.SetDefault("ldapmaxconcurrency", 10)

	viper.SetConfigName("cboxgroupd")
	viper.AddConfigPath("/etc/cboxgroupd/")

	flag.BoolVar(&fVersion, "version", false, "Show version")
	flag.Int("port", 2002, "Port to listen for connections")
	flag.String("ldaphostname", "xldap.cern.ch", "Hostname of the LDAP server")
	flag.Int("ldapport", 389, "Port of LDAP server")
	flag.Uint("ldappagelimit", 1000, "Page limit for paged searchs")
	flag.String("redishostname", "localhost", "Hostname of the Redis server")
	flag.Int("redisport", 6379, "Port of Redis server")
	flag.Int("redisdb", 0, "Redis number database for keys isolation (0-15)")
	flag.Int("redisttl", 60, "Number of seconds to expire cached entries in Redis")
	flag.String("applog", "stderr", "File to log application data")
	flag.String("httplog", "stderr", "File to log HTTP requests")
	flag.String("secret", "changeme!!!", "Share secret between services to authenticate requests")
	flag.Int("ldapmaxconcurrency", 100, "Number of concurrent connections to LDAP for update operations")
	flag.String("config", "", "Configuration file to use")

	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
	pflag.Parse()
	viper.BindPFlags(pflag.CommandLine)
}

func main() {

	if fVersion {
		showVersion()
	}

	if viper.GetString("config") != "" {
		viper.SetConfigFile(viper.GetString("config"))
	}

	err := viper.ReadInConfig()
	if err != nil {
		panic(fmt.Errorf("Fatal error config file: %s \n", err))
	}

	config := zap.NewProductionConfig()
	config.OutputPaths = []string{viper.GetString("applog")}
	logger, _ := config.Build()

	lgl := ldapgrouplooker.New(viper.GetString("ldaphostname"), viper.GetInt("ldapport"), uint32(viper.GetInt("ldappagelimit")))
	rgl := redisgrouplooker.New(viper.GetString("redishostname"), viper.GetInt("redisport"), viper.GetInt("redisdb"), viper.GetInt("redisttl"), lgl)

	router := mux.NewRouter()

	protectedUsersInGroup := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.UsersInGroup(logger, rgl))
	protectedUsersInComputingGroup := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.UsersInComputingGroup(logger, rgl))
	protectedUserGroups := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.UserGroups(logger, rgl))
	protectedUserComputingGroups := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.UserComputingGroups(logger, rgl))
	protectedUsersInGroupTTL := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.UsersInGroupTTL(logger, rgl))
	protectedUsersInComputingGroupTTL := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.UsersInComputingGroupTTL(logger, rgl))
	protectedUserGroupsTTL := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.UserGroupsTTL(logger, rgl))
	protectedUserComputingGroupsTTL := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.UserComputingGroupsTTL(logger, rgl))

	protectedUpdateUsersInGroup := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.UpdateUsersInGroup(logger, rgl, viper.GetInt("ldapmaxconcurrency")))
	protectedUpdateUserGroups := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.UpdateUserGroups(logger, rgl, viper.GetInt("ldapmaxconcurrency")))

	protectedSearch := handlers.CheckSharedSecret(logger, viper.GetString("secret"), handlers.Search(logger, rgl))

	router.Handle("/api/v1/membership/usersingroup/{gid}", protectedUsersInGroup).Methods("GET")
	router.Handle("/api/v1/membership/usersincomputinggroup/{gid}", protectedUsersInComputingGroup).Methods("GET")
	router.Handle("/api/v1/membership/usergroups/{uid}", protectedUserGroups).Methods("GET")
	router.Handle("/api/v1/membership/usercomputinggroups/{uid}", protectedUserComputingGroups).Methods("GET")

	router.Handle("/api/v1/membership/usersingroupttl/{gid}", protectedUsersInGroupTTL).Methods("GET")
	router.Handle("/api/v1/membership/usersincomputinggroupttl/{gid}", protectedUsersInComputingGroupTTL).Methods("GET")
	router.Handle("/api/v1/membership/usergroupsttl/{uid}", protectedUserGroupsTTL).Methods("GET")
	router.Handle("/api/v1/membership/usercomputinggroupsttl/{uid}", protectedUserComputingGroupsTTL).Methods("GET")

	router.Handle("/api/v1/update/usersingroup", protectedUpdateUsersInGroup).Methods("POST")
	router.Handle("/api/v1/update/usergroups", protectedUpdateUserGroups).Methods("POST")

	router.Handle("/api/v1/search/{filter}", protectedSearch).Methods("GET")

	out := getHTTPLoggerOut(viper.GetString("httplog"))
	loggedRouter := gh.LoggingHandler(out, router)

	logger.Info("server is listening", zap.Int("port", viper.GetInt("port")))
	logger.Warn("server stopped", zap.Error(http.ListenAndServe(fmt.Sprintf(":%d", viper.GetInt("port")), loggedRouter)))
}

func getHTTPLoggerOut(filename string) *os.File {
	if filename == "stderr" {
		return os.Stderr
	} else if filename == "stdout" {
		return os.Stdout
	} else {
		fd, err := os.OpenFile(filename, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
		if err != nil {
			log.Fatal(err)
		}
		return fd
	}
}

func showVersion() {
	// if gitTag is not empty we are on release build
	if gitTag != "" {
		fmt.Printf("%s %s commit:%s release-build\n", appName, gitNearestTag, gitCommit)
		os.Exit(0)
	}
	fmt.Printf("%s %s commit:%s dev-build\n", appName, gitNearestTag, gitCommit)
	os.Exit(0)
}
